library/core/src/iter/traits/iterator.rs

```diff
@@ -3011,6 +3015,18 @@ pub trait Iterator {
     /// assert_eq!(v_cloned, vec![1, 2, 3]);
     /// assert_eq!(v_map, vec![1, 2, 3]);
     /// ```
+    ///
+    /// To get the best performance, try to clone late:
```

Comment:
Adding a Clippy lint that suggests to move the clone() later in the chain sounds tricky.
Adding compiler optimizations for such case sounds trickier.

Reply:
The lint is already in a PR, only its docs need some ❤️. I'll see when I get to suggest something.

Reply:
Update: The lint has been merged.